src/subversion_rando/goal.py

```diff
@@ -0,0 +1,244 @@
+
+
+def WriteLogs(romWriter: RomWriter, address: int, goals: list[Event]) -> None:
+    hint_address = address + 2 + (2 * len(goals))
+
+    for i, goal in enumerate(goals):
+        romWriter.writeBytes(address, struct.pack('<H', GetShortAddress(hint_address)))
+        hint_address = WriteLogEntry(romWriter, hint_address, i, *goal)
+        address += 2
+    romWriter.writeBytes(address, struct.pack('<H', 0))
+
+
+def WriteMessageBoxes(romWriter: RomWriter, address: int, goals: list[Event]) -> None:
+    message_address = address + (6 * (len(goals) + 2))
+
+    message = ConvertToMessagebox('OBJECTIVES NOT COMPLETE')
+    romWriter.writeBytes(address, struct.pack('<HHH', 0x8436, 0x825A, GetShortAddress(message_address)))
+    romWriter.writeBytes(message_address, message)
+    address += 6
+    message_address += len(message)
+    for i,goal in enumerate(goals):
+        message = ConvertToMessagebox(f'GOAL {chr(ord("A") + i)}: {goal[1]}')
+        romWriter.writeBytes(address, struct.pack('<HHH', 0x8436, 0x825A, GetShortAddress(message_address)))
+        romWriter.writeBytes(message_address, message)
+        address += 6
+        message_address += len(message)
+    romWriter.writeBytes(address, struct.pack('<HHH', 0x8436, 0x825A, GetShortAddress(message_address)))
+
+
+def generate_goals(game: Game, romWriter: RomWriter) -> None:
+    count = game.options.objective_rando
+    if count > max_goal_count:
+        raise Exception(f'Cannot generate {count} goals, which is more than the maximum allowed of {max_goal_count}.')
+
+    valid_events = list(events)
+    if game.options.area_rando:
+        bad_events = ["POWER OFF"]
+        valid_events = [
+            [sub_event for sub_event in event if sub_event[1] not in bad_events] 
+            for event in valid_events
+        ]
+    valid_events = [event for event in valid_events if event]
+
+    # select goals
+    goals = [random.sample(subgoals, 1)[0] for subgoals in random.sample(valid_events, count)]
+
+    # write room init function to check state conditions
+    romWriter.writeBytes(final_room_init_address, struct.pack('<H', check_event_func))
+
+    # write event list
+    data = bytes([goal[0] for goal in goals])
+    romWriter.writeBytes(event_address, data)
+
+    WriteLogs(romWriter, log_mission_address, goals)
+    WriteMessageBoxes(romWriter, messages_boxes_address + (base_message_id * 6), goals)
+
+    # write out map station plm data
+    remaining_map_stations = list(map_stations)
+    random.shuffle(remaining_map_stations)
+    for i, map_station in enumerate(remaining_map_stations):
+        i = i % len(goals)
+        goal = goals[i]
+
+        # change map station into message station
+        romWriter.writeBytes(map_station + 0, struct.pack('<H', 0xF2C0))
+        # skip over the 2 bytes for the plm position
+        # write message and event into plm arg
+        romWriter.writeBytes(map_station + 4, struct.pack('BB', base_event_id + i, base_message_id + i + 1))
```

Comment:
I'm trying to understand what's going on in this for loop.

It looks like it's iterating through all of the map stations, right? (even though the name `remaining_map_stations` makes it seem like it's not all of them)

And it pairs each map station with a goal. But what I found in my test playthrough is that not every map station gives me something.
Is it that only the first map station associated with 1 specific goal gives me the message box?
If someone chooses 1 as the number of objectives, will every map station be a source of information for that objective?
But only the 1st one I visit will give me the information?

Reply:
The name `remaining_map_stations` is probably bad, but I needed a duplicate of the underlying list so I can safely modify it without affecting subsequent runs.

So if you collect a hint from a map station, that sets an event which shows up in the logbook. The map stations are no longer interactable if the the event is set. So if multiple map stations share the same hint, after you use one of them, the other duplicates don't work anymore. The player can at a glance know to not walk up to the map station because the animation for a "read" station is different. This does mean if you have only 1 goal, then all stations share the same hint, and you'll only receive it the one time.